lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,332 @@
+
+/**
+ * Module dependencies.
+ */
+
+var _ = require('lodash');
+var AccessDeniedError = require('../errors/access-denied-error');
+var AuthenticateHandler = require('../handlers/authenticate-handler');
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidClientError = require('../errors/invalid-client-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var InvalidScopeError = require('../errors/invalid-scope-error');
+var OAuthError = require('../errors/oauth-error');
+var Promise = require('bluebird');
+var Request = require('../request');
+var Response = require('../response');
+var ServerError = require('../errors/server-error');
+var UnauthorizedClientError = require('../errors/unauthorized-client-error');
+var is = require('../validator/is');
+var tokenUtil = require('../utils/token-util');
+var url = require('url');
+
+/**
+ * Response types.
+ */
+
+var responseTypes = {
+  code: require('../response-types/code-response-type'),
+  token: require('../response-types/token-response-type')
+};
+
+/**
+ * Constructor.
+ */
+
+function AuthorizeHandler(options) {
+  options = options || {};
+
+  if (options.authenticateHandler && !options.authenticateHandler.handle) {
+    throw new InvalidArgumentError('Invalid argument: authenticateHandler does not implement `handle()`');
+  }
+
+  if (!options.authorizationCodeLifetime) {
+    throw new InvalidArgumentError('Missing parameter: `authorizationCodeLifetime`');
+  }
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getClient) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getClient()`');
+  }
+
+  if (!options.model.saveAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveAuthorizationCode()`');
+  }
+
+  this.authorizationCodeLifetime = options.authorizationCodeLifetime;
+  this.authenticateHandler = options.authenticateHandler || new AuthenticateHandler(options);
+  this.model = options.model;
+}
+
+/**
+ * Authorize Handler.
+ */
+
+AuthorizeHandler.prototype.handle = function(request, response) {
+  if (!(request instanceof Request)) {
+    throw new InvalidArgumentError('Invalid argument: `request` must be an instance of Request');
+  }
+
+  if (!(response instanceof Response)) {
+    throw new InvalidArgumentError('Invalid argument: `response` must be an instance of Response');
+  }
+
+  if (false === request.body.allowed) {
```

Comment:
Indeed. As you can read in that comment, I struggled a bit with this one.

The `allowed` parameter is a feature that I honestly would not have implemented if I started writing an oauth server from scratch.

To anyone just reading this, I'll explain the issue:

When a user is faced with a common authorization page such as:

    Would you like to allow <My App> access to your personal information?

    [ Allow ] [ Deny ]

Of course, by clicking the "Allow" button, you'll continue through the authorization flow, but when clicking the "Deny" button, you don't really need to make an HTTP request. You can, in fact, just close the webpage and get the same result.

So, really, this `allowed` feature only exists for developers to continue through authorization flow but have it purposefully "fail".

@night: I'll change it to string `"false"` because I agree that JSON shouldn't be a requirement but I'd prefer not adding a model method for this feature because, as I said above, I don't really see it as a very interesting _model- feature.